actor/src/main/scala/org/apache/pekko/util/RandomShortProvider.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * license agreements; and to You under the Apache License, version 2.0:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * This file is part of the Apache Pekko project, derived from Akka.
+ */
+
+package org.apache.pekko.util
+
+import org.apache.pekko.annotation.InternalApi
+
+import java.util.Random
+
+/**
+ * An efficient thread safe generator of pseudo random shorts based on
+ * https://en.wikipedia.org/wiki/Double_hashing#Enhanced_double_hashing.
+ *
+ * Note that due to the usage of synchronized this method is optimized
+ * for the happy case (i.e. least contention) on multiple threads.
+ */
+
+@InternalApi
+private[pekko] class RandomShortProvider(seed: Random) {
+  // These methods are volatile so that they stay within cache
+  private var index = seed.nextLong
+  private var increment = seed.nextLong
+  private var count = 1L
+
+  def nextId(): Short = synchronized {
+    val result = (0xFFFFFFFF & index).asInstanceOf[Short]
```

Comment:
would it be possible to consider https://github.com/mdedetrich/incubator-pekko/pull/1 ? @alexandru suggested that 'synchronized' is a bad solution and the semaphore can timeout (while synchronized can just hang).

Reply:
Sure ill look at it tomorrow, will be busy tonight

Reply:
So I have just update the version to use a semaphore as well as making the timeout configurable plus I did some other improvements to make the config reading/parsing consistent with the rest of Pekko.

I used the exact same version that you suggested (with a minor performance improvement to avoid creating a new class in the failure case of acquiring a lock) and the performance nosedived.

```
[info] IdGeneratorBanchmark.measureEnhancedDoubleHash                 thrpt    3  ≈ 10⁻¹⁰            ops/ns
[info] IdGeneratorBanchmark.measureSecureRandom                       thrpt    3    0,001 ±   0,001  ops/ns
[info] IdGeneratorBanchmark.measureThreadLocalRandom                  thrpt    3    0,311 ±   0,031  ops/ns
[info] IdGeneratorBanchmark.multipleThreadsMeasureEnhancedDoubleHash  thrpt    3  ≈ 10⁻¹⁰            ops/ns
[info] IdGeneratorBanchmark.multipleThreadsMeasureSecureRandom        thrpt    3    0,001 ±   0,001  ops/ns
[info] IdGeneratorBanchmark.multipleThreadsMeasureThreadLocalRandom   thrpt    3    0,608 ±   0,036  ops/ns
```

Reply:
So using a much shorter timeout the improvements are better (i.e. this is the results with 10 nanos)

```
[info] IdGeneratorBanchmark.measureEnhancedDoubleHash                 thrpt    3  0,001 ±  0,001  ops/ns
[info] IdGeneratorBanchmark.measureSecureRandom                       thrpt    3  0,001 ±  0,001  ops/ns
[info] IdGeneratorBanchmark.measureThreadLocalRandom                  thrpt    3  0,312 ±  0,074  ops/ns
[info] IdGeneratorBanchmark.multipleThreadsMeasureEnhancedDoubleHash  thrpt    3  0,001 ±  0,001  ops/ns
[info] IdGeneratorBanchmark.multipleThreadsMeasureSecureRandom        thrpt    3  0,001 ±  0,001  ops/ns
[info] IdGeneratorBanchmark.multipleThreadsMeasureThreadLocalRandom   thrpt    3  0,622 ±  0,018  ops/ns
```

I am not sure what problem we are trying to achieve with a semaphore though, in such a hot loop with only a single permit its actually slower than `synchronized` which because the JVM know its only single threaded (i.e. permit of 1)  is hyper optimized for this case.

Reply:
@pjfanning You just approved PR, did you read this?